File: C++/06-Stack-Queue/stack_queue.cpp
// 06 - Stack and Queue: both from scratch, plus the monotonic-stack patterns
// that show up constantly in interviews.
//
// Build & run:
//   g++ -std=c++17 -O2 -Wall stack_queue.cpp -o stack_queue && ./stack_queue

#include <cassert>
#include <deque>
#include <functional>
#include <iostream>
#include <stack>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

// ============================================================================
// 1. Stack over a vector (LIFO)
// ============================================================================
template <typename T>
class ArrayStack {
   public:
    void push(const T& value) { data_.push_back(value); }   // amortised O(1)

    T pop() {                                                // O(1)
        if (data_.empty()) throw out_of_range("pop from empty stack");
        T value = data_.back();
        data_.pop_back();
        return value;
    }

    const T& top() const {
        if (data_.empty()) throw out_of_range("top of empty stack");
        return data_.back();
    }

    bool empty() const { return data_.empty(); }
    size_t size() const { return data_.size(); }

   private:
    vector<T> data_;
};

// ============================================================================
// 2. Circular-buffer queue (FIFO)
// ============================================================================
// head_ is where the next dequeue happens; the write position is derived as
// (head_ + count_) % capacity. Nothing is ever shifted, so both ends are O(1).
// A separate count_ distinguishes full from empty, which head == tail cannot.
template <typename T>
class CircularQueue {
   public:
    explicit CircularQueue(size_t capacity)
        : buf_(capacity), capacity_(capacity) {
        if (capacity == 0) throw invalid_argument("capacity must be positive");
    }

    void enqueue(const T& value) {
        if (full()) throw overflow_error("queue is full");
        buf_[(head_ + count_) % capacity_] = value;
        count_++;
    }

    T dequeue() {
        if (empty()) throw out_of_range("dequeue from empty queue");
        T value = buf_[head_];
        head_ = (head_ + 1) % capacity_;
        count_--;
        return value;
    }

    const T& front() const {
        if (empty()) throw out_of_range("front of empty queue");
        return buf_[head_];
    }

    bool empty() const { return count_ == 0; }
    bool full() const { return count_ == capacity_; }
    size_t size() const { return count_; }

    vector<T> toVector() const {
        vector<T> out;
        out.reserve(count_);
        for (size_t i = 0; i < count_; i++)
            out.push_back(buf_[(head_ + i) % capacity_]);
        return out;
    }

   private:
    vector<T> buf_;
    size_t capacity_;
    size_t head_ = 0;
    size_t count_ = 0;
};

// ============================================================================
// 3. MinStack - O(1) minimum
// ============================================================================
// Storing (value, min_so_far) pairs trades O(n) space for an O(1) query;
// scanning for the minimum on demand would be O(n) per call.
class MinStack {
   public:
    void push(int val) {
        int currentMin = data_.empty() ? val : min(val, data_.back().second);
        data_.push_back({val, currentMin});
    }

    int pop() {
        if (data_.empty()) throw out_of_range("pop from empty stack");
        int value = data_.back().first;
        data_.pop_back();
        return value;
    }

    int top() const { return data_.back().first; }
    int getMin() const { return data_.back().second; }
    bool empty() const { return data_.empty(); }

   private:
    vector<pair<int, int>> data_;
};

// ============================================================================
// 4. Queue built from two stacks
// ============================================================================
// Pouring `in_` into `out_` reverses the order, so the oldest element ends up
// on top of `out_`. Each element moves at most twice: amortised O(1).
class QueueViaStacks {
   public:
    void push(int value) { in_.push(value); }

    int pop() {
        shift();
        if (out_.empty()) throw out_of_range("pop from empty queue");
        int value = out_.top();
        out_.pop();
        return value;
    }

    int front() {
        shift();
        if (out_.empty()) throw out_of_range("front of empty queue");
        return out_.top();
    }

    size_t size() const { return in_.size() + out_.size(); }

   private:
    void shift() {
        if (out_.empty()) {              // only pour when it has run dry
            while (!in_.empty()) {
                out_.push(in_.top());
                in_.pop();
            }
        }
    }

    stack<int> in_, out_;
};

// ============================================================================
// 5. Matching / nesting
// ============================================================================
// A closer must match the most recent opener - exactly what a stack tracks.
bool isBalanced(const string& s) {
    unordered_map<char, char> pairs{{')', '('}, {']', '['}, {'}', '{'}};
    vector<char> st;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            st.push_back(c);
        } else if (pairs.count(c)) {
            if (st.empty() || st.back() != pairs[c]) return false;
            st.pop_back();
        }
    }
    return st.empty();                   // leftovers mean unclosed openers
}

// ============================================================================
// 6. Monotonic stack
// ============================================================================
// Next strictly greater element to the right, or -1. O(n): every index is
// pushed once and popped at most once.
vector<int> nextGreater(const vector<int>& nums) {
    vector<int> result(nums.size(), -1);
    vector<int> st;                      // indices; their values decrease
    for (size_t i = 0; i < nums.size(); i++) {
        while (!st.empty() && nums[st.back()] < nums[i]) {
            result[st.back()] = nums[i]; // nums[i] is the answer for that index
            st.pop_back();
        }
        st.push_back(int(i));
    }
    return result;
}

// Days to wait for a warmer temperature. Same pattern, distances instead of
// values - which is why the stack holds indices.
vector<int> dailyTemperatures(const vector<int>& temps) {
    vector<int> result(temps.size(), 0);
    vector<int> st;
    for (int i = 0; i < int(temps.size()); i++) {
        while (!st.empty() && temps[st.back()] < temps[i]) {
            result[st.back()] = i - st.back();
            st.pop_back();
        }
        st.push_back(i);
    }
    return result;
}

// Largest rectangle in a histogram. O(n) with a monotonic increasing stack.
// When a shorter bar arrives, every taller bar on the stack can no longer
// extend right, so its maximal rectangle is finalised. The trailing 0 sentinel
// flushes whatever remains.
long long largestRectangle(vector<int> heights) {
    heights.push_back(0);                // sentinel
    vector<int> st;
    long long best = 0;
    for (int i = 0; i < int(heights.size()); i++) {
        while (!st.empty() && heights[st.back()] >= heights[i]) {
            long long height = heights[st.back()];
            st.pop_back();
            long long left = st.empty() ? 0 : st.back() + 1;
            best = max(best, height * (i - left));
        }
        st.push_back(i);
    }
    return best;
}

// ============================================================================
// 7. Simulation
// ============================================================================
// Reverse Polish notation. Operands wait on the stack until an operator claims
// the last two - and the order of those two matters for - and /.
int evalRPN(const vector<string>& tokens) {
    vector<int> st;
    for (const string& token : tokens) {
        if (token == "+" || token == "-" || token == "*" || token == "/") {
            int b = st.back(); st.pop_back();
            int a = st.back(); st.pop_back();
            if (token == "+") st.push_back(a + b);
            else if (token == "-") st.push_back(a - b);
            else if (token == "*") st.push_back(a * b);
            else st.push_back(a / b);    // C++ division truncates toward zero
        } else {
            st.push_back(stoi(token));
        }
    }
    return st.back();
}

// Maximum of every window of size k. O(n) with a monotonic deque: the front is
// always the window maximum, and any value smaller than the arriving one can
// never win again.
vector<int> slidingWindowMax(const vector<int>& nums, size_t k) {
    if (k == 0 || k > nums.size()) throw invalid_argument("bad window size");
    deque<size_t> dq;                    // indices, values decreasing
    vector<int> out;
    out.reserve(nums.size() - k + 1);
    for (size_t i = 0; i < nums.size(); i++) {
        while (!dq.empty() && dq.front() + k <= i) dq.pop_front();  // expired
        while (!dq.empty() && nums[dq.back()] <= nums[i]) dq.pop_back();
        dq.push_back(i);
        if (i + 1 >= k) out.push_back(nums[dq.front()]);
    }
    return out;
}

// ============================================================================
// demo
// ============================================================================
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    ArrayStack<int> st;
    st.push(1);
    st.push(2);
    assert(st.top() == 2 && st.size() == 2);
    assert(st.pop() == 2 && st.pop() == 1 && st.empty());
    bool threw = false;
    try { st.pop(); } catch (const out_of_range&) { threw = true; }
    assert(threw);

    CircularQueue<int> q(3);
    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    assert(q.full() && (q.toVector() == vector<int>{1, 2, 3}));
    assert(q.dequeue() == 1);
    q.enqueue(4);                        // wraps around the buffer
    assert((q.toVector() == vector<int>{2, 3, 4}));
    assert(q.front() == 2 && q.size() == 3);

    MinStack ms;
    for (int v : {5, 3, 7, 3}) ms.push(v);
    assert(ms.getMin() == 3 && ms.top() == 3);
    ms.pop();
    assert(ms.getMin() == 3);
    ms.pop();                            // removes 7
    ms.pop();                            // removes the first 3
    assert(ms.getMin() == 5);

    QueueViaStacks qs;
    for (int v : {1, 2, 3}) qs.push(v);
    assert(qs.front() == 1 && qs.pop() == 1);
    qs.push(4);
    assert(qs.pop() == 2 && qs.pop() == 3 && qs.pop() == 4);

    assert(isBalanced("({[]})"));
    assert(isBalanced(""));
    assert(!isBalanced("(]"));
    assert(!isBalanced("(("));

    assert((nextGreater({2, 1, 2, 4, 3}) == vector<int>{4, 2, 4, -1, -1}));
    assert((dailyTemperatures({73, 74, 75, 71, 69, 72, 76, 73}) ==
            vector<int>{1, 1, 4, 2, 1, 1, 0, 0}));
    assert(largestRectangle({2, 1, 5, 6, 2, 3}) == 10);
    assert(largestRectangle({2, 2}) == 4);

    assert(evalRPN({"2", "1", "+", "3", "*"}) == 9);
    assert(evalRPN({"4", "13", "5", "/", "+"}) == 6);

    assert((slidingWindowMax({1, 3, -1, -3, 5, 3, 6, 7}, 3) ==
            vector<int>{3, 3, 5, 5, 6, 7}));

    cout << "06-Stack-Queue (C++): all checks passed\n";
    return 0;
}

File: Python/06-Stack-Queue/stack_queue.py
"""
06 - Stack and Queue: both structures from scratch, plus the monotonic-stack
patterns that show up constantly in interviews.

Run:  python stack_queue.py
"""

from __future__ import annotations

from collections import deque
from typing import Any


# ============================================================================
# 1. Stack (LIFO)
# ============================================================================
class Stack:
    """A list is already a perfect stack; this wraps it in the vocabulary.

    push/pop/peek are all O(1).
    """

    def __init__(self) -> None:
        self._items: list[Any] = []

    def push(self, item: Any) -> None:
        self._items.append(item)

    def pop(self) -> Any:
        if not self._items:
            raise IndexError("pop from empty stack")
        return self._items.pop()

    def peek(self) -> Any:
        if not self._items:
            raise IndexError("peek at empty stack")
        return self._items[-1]

    def is_empty(self) -> bool:
        return not self._items

    def __len__(self) -> int:
        return len(self._items)

    def __repr__(self) -> str:
        return f"Stack({self._items})  <- top"


# ============================================================================
# 2. Queue (FIFO) - circular buffer over a fixed array
# ============================================================================
class CircularQueue:
    """Fixed-capacity FIFO queue with O(1) enqueue and dequeue.

    head is where the next dequeue happens; tail is where the next enqueue
    goes. Both wrap with % capacity, so no element is ever shifted.

    A separate `count` distinguishes full from empty - otherwise head == tail
    would be ambiguous.
    """

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._buf: list[Any] = [None] * capacity
        self._capacity = capacity
        self._head = 0
        self._count = 0

    def enqueue(self, item: Any) -> None:
        if self.is_full():
            raise OverflowError("queue is full")
        tail = (self._head + self._count) % self._capacity
        self._buf[tail] = item
        self._count += 1

    def dequeue(self) -> Any:
        if self.is_empty():
            raise IndexError("dequeue from empty queue")
        item = self._buf[self._head]
        self._buf[self._head] = None           # release the reference
        self._head = (self._head + 1) % self._capacity
        self._count -= 1
        return item

    def front(self) -> Any:
        if self.is_empty():
            raise IndexError("front of empty queue")
        return self._buf[self._head]

    def is_empty(self) -> bool:
        return self._count == 0

    def is_full(self) -> bool:
        return self._count == self._capacity

    def __len__(self) -> int:
        return self._count

    def to_list(self) -> list[Any]:
        return [self._buf[(self._head + i) % self._capacity]
                for i in range(self._count)]


# ============================================================================
# 3. MinStack - O(1) minimum
# ============================================================================
class MinStack:
    """Stack that also reports its minimum in O(1).

    Trick: store (value, min_so_far) pairs. The extra O(n) space buys the O(1)
    query - scanning for the min on demand would be O(n) per call.
    """

    def __init__(self) -> None:
        self._stack: list[tuple[int, int]] = []

    def push(self, val: int) -> None:
        current_min = val if not self._stack else min(val, self._stack[-1][1])
        self._stack.append((val, current_min))

    def pop(self) -> int:
        if not self._stack:
            raise IndexError("pop from empty stack")
        return self._stack.pop()[0]

    def top(self) -> int:
        return self._stack[-1][0]

    def min(self) -> int:
        return self._stack[-1][1]

    def __len__(self) -> int:
        return len(self._stack)


# ============================================================================
# 4. Queue built from two stacks
# ============================================================================
class QueueViaStacks:
    """FIFO using two LIFOs. Amortised O(1) per operation.

    Pouring `inbox` into `outbox` reverses the order, so the oldest element
    ends up on top of `outbox`. Each element is moved at most twice in its
    lifetime, which is what makes the average constant.
    """

    def __init__(self) -> None:
        self._inbox: list[Any] = []
        self._outbox: list[Any] = []

    def enqueue(self, item: Any) -> None:
        self._inbox.append(item)               # always O(1)

    def dequeue(self) -> Any:
        self._shift()
        if not self._outbox:
            raise IndexError("dequeue from empty queue")
        return self._outbox.pop()

    def front(self) -> Any:
        self._shift()
        if not self._outbox:
            raise IndexError("front of empty queue")
        return self._outbox[-1]

    def _shift(self) -> None:
        if not self._outbox:                   # only pour when it runs dry
            while self._inbox:
                self._outbox.append(self._inbox.pop())

    def __len__(self) -> int:
        return len(self._inbox) + len(self._outbox)


# ============================================================================
# 5. Matching / nesting
# ============================================================================
def is_balanced(s: str) -> bool:
    """Valid parentheses. O(n) time, O(n) space.

    A closer must match the most recent opener - that "most recent" is exactly
    what a stack gives you.
    """
    pairs = {")": "(", "]": "[", "}": "{"}
    stack: list[str] = []
    for ch in s:
        if ch in "([{":
            stack.append(ch)
        elif ch in pairs:
            if not stack or stack.pop() != pairs[ch]:
                return False
    return not stack                           # leftovers mean unclosed openers


# ============================================================================
# 6. Monotonic stack
# ============================================================================
def next_greater(nums: list[int]) -> list[int]:
    """For each element, the next strictly greater element to its right, or -1.

    O(n): each index is pushed once and popped at most once, so the inner
    while loop is amortised constant.
    """
    result = [-1] * len(nums)
    stack: list[int] = []                      # indices, values decreasing
    for i, x in enumerate(nums):
        while stack and nums[stack[-1]] < x:
            result[stack.pop()] = x            # x is the answer for that index
        stack.append(i)
    return result


def daily_temperatures(temps: list[int]) -> list[int]:
    """Days to wait for a warmer temperature. Same pattern, distances instead.

    This is why the stack holds INDICES: we need i - j, not the value.
    """
    result = [0] * len(temps)
    stack: list[int] = []
    for i, t in enumerate(temps):
        while stack and temps[stack[-1]] < t:
            j = stack.pop()
            result[j] = i - j
        stack.append(i)
    return result


def largest_rectangle(heights: list[int]) -> int:
    """Largest rectangle in a histogram. O(n) with a monotonic increasing stack.

    When a bar shorter than the stack top arrives, the popped bar can no longer
    extend right - so its maximal rectangle is finalised here. The sentinel 0
    at the end flushes whatever is left on the stack.
    """
    stack: list[int] = []                      # indices, heights increasing
    best = 0
    for i, h in enumerate(heights + [0]):      # sentinel forces a full flush
        while stack and heights[stack[-1]] >= h:
            height = heights[stack.pop()]
            left = stack[-1] + 1 if stack else 0
            best = max(best, height * (i - left))
        stack.append(i)
    return best


# ============================================================================
# 7. Simulation
# ============================================================================
def eval_rpn(tokens: list[str]) -> int:
    """Evaluate reverse Polish notation. O(n) time and space.

    Operands wait on the stack until an operator claims the last two.
    """
    stack: list[int] = []
    ops = {
        "+": lambda a, b: a + b,
        "-": lambda a, b: a - b,
        "*": lambda a, b: a * b,
        "/": lambda a, b: int(a / b),          # truncates toward zero
    }
    for token in tokens:
        if token in ops:
            b = stack.pop()
            a = stack.pop()                    # order matters for - and /
            stack.append(ops[token](a, b))
        else:
            stack.append(int(token))
    return stack.pop()


def simplify_path(path: str) -> str:
    """Unix path canonicalisation: '/a/./b/../c' -> '/a/c'. O(n)."""
    stack: list[str] = []
    for part in path.split("/"):
        if part == "..":
            if stack:
                stack.pop()                    # go up one directory
        elif part and part != ".":
            stack.append(part)
    return "/" + "/".join(stack)


def sliding_window_max(nums: list[int], k: int) -> list[int]:
    """Maximum of every window of size k. O(n) with a monotonic deque.

    The deque holds indices whose values decrease. The front is always the
    window maximum; anything smaller arriving from the right makes older,
    smaller values permanently useless.
    """
    if k <= 0 or k > len(nums):
        raise ValueError("bad window size")
    dq: deque[int] = deque()
    out: list[int] = []
    for i, x in enumerate(nums):
        while dq and dq[0] <= i - k:           # drop indices that left the window
            dq.popleft()
        while dq and nums[dq[-1]] <= x:        # drop values that can never win
            dq.pop()
        dq.append(i)
        if i >= k - 1:
            out.append(nums[dq[0]])
    return out


# ============================================================================
# demo
# ============================================================================
def demo() -> None:
    st = Stack()
    st.push(1)
    st.push(2)
    assert st.peek() == 2 and len(st) == 2
    assert st.pop() == 2 and st.pop() == 1 and st.is_empty()
    try:
        st.pop()
        raise AssertionError("expected IndexError")
    except IndexError:
        pass

    q = CircularQueue(3)
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    assert q.is_full() and q.to_list() == [1, 2, 3]
    assert q.dequeue() == 1
    q.enqueue(4)                               # wraps around the buffer
    assert q.to_list() == [2, 3, 4]
    assert q.front() == 2 and len(q) == 3

    ms = MinStack()
    for v in (5, 3, 7, 3):
        ms.push(v)
    assert ms.min() == 3 and ms.top() == 3
    ms.pop()
    assert ms.min() == 3
    ms.pop()                                   # removes the 7
    ms.pop()                                   # removes the first 3
    assert ms.min() == 5

    qs = QueueViaStacks()
    for v in (1, 2, 3):
        qs.enqueue(v)
    assert qs.front() == 1 and qs.dequeue() == 1
    qs.enqueue(4)
    assert [qs.dequeue() for _ in range(3)] == [2, 3, 4]

    assert is_balanced("({[]})")
    assert is_balanced("")
    assert not is_balanced("(]")
    assert not is_balanced("((")

    assert next_greater([2, 1, 2, 4, 3]) == [4, 2, 4, -1, -1]
    assert daily_temperatures([73, 74, 75, 71, 69, 72, 76, 73]) == [1, 1, 4, 2, 1, 1, 0, 0]
    assert largest_rectangle([2, 1, 5, 6, 2, 3]) == 10
    assert largest_rectangle([2, 2]) == 4

    assert eval_rpn(["2", "1", "+", "3", "*"]) == 9
    assert eval_rpn(["4", "13", "5", "/", "+"]) == 6

    assert simplify_path("/a/./b/../../c/") == "/c"
    assert simplify_path("/../") == "/"

    assert sliding_window_max([1, 3, -1, -3, 5, 3, 6, 7], 3) == [3, 3, 5, 5, 6, 7]

    print("06-Stack-Queue (Python): all checks passed")


if __name__ == "__main__":
    demo()
